### PR TITLE
fix(windows): Settings refresh management

### LINKED ISSUE #4011: feat(windows): Keyman Configuration should refresh if changes are made by other processes
**Is your feature request related to a problem? Please describe.**

Keyman Configuration should refresh if changes are made by other processes. For example, if you double-click to install a package, that should refresh another instance of Keyman Configuration currently running; or if you upgrade a package, etc.

---

**Keyman for Windows/macOS/Linux:**
 - OS: Windows 10 x64 1909 
 - Keyman version: 14.0.192-alpha

### LINKED ISSUE #4039: bug(windows): uninstalling then reinstalling a package fails in a single session of Keyman Configuration
**Describe the bug**

If you attempt to uninstall and then reinstall a package within a single session of Keyman Configuration, it thinks the package is still present and then fails to uninstall it (for obvious reasons!)

**To Reproduce**

1. Uninstall a package.
2. Reinstall it.

---

**Keyman for Windows/macOS/Linux:**
 - OS: Windows 10 2004 x64 
 - Keyman version: 14.0.194-alpha

### LINKED ISSUE #4119: bug(windows): set a hotkey after a fresh install
**Describe the bug**

Setting a hotkey after a fresh install does not work as expected. The hotkey would get set for a different keyboard and not the one intended by the user.

**To Reproduce**

1. Install the Keyman 14.0.204-alpha and Go to 'Configuration'
2. Click on 'Hotkeys'
3. Scroll down to set a hotkey for a keyboard and take note of which keyboard you set the hotkey for and click "OK"

**Expected behavior**

The hotkey should be for the intended keyboard, not any other one which was not set for.

**Diag report**

@mcdurdin 
Here is the diag report: https://drive.google.com/file/d/1bYbHwbH7V3vjxuUp5YTIHmwvODZ_SirX/view?usp=sharing.

---

**Keyman for Windows:**
 - OS: Windows 10 Pro 20H2 x64
 - Keyman version: 14.0.204-alpha
 - Target Application: Keyman

### LINKED ISSUE #4121: bug (windows): changes do not apply immediately
**Describe the bug**

Two keyboards were installed after a fresh install, but the changes do not apply immediately
image

**Expected behavior**

Any keyboards installed should be considered as a change which is expected to be applied immediately.

---

**Keyman for Windows:**
 - OS: Windows 10 Pro 1909 x64
 - Keyman version: 4.0.204-alpha
 - Target Application: Keyman Configuration

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #4011.
Fixes #4039.
Fixes #4121.
Fixes #4119.
Fixes KEYMAN-WINDOWS-6A.
Fixes KEYMAN-WINDOWS-6B.
Fixes KEYMAN-WINDOWS-4B.
Fixes KEYMAN-WINDOWS-4C.

I am sorry that this touches as many files as it does. Hopefully, most of the changes are relatively simple and clear.

This change completes the immediate-change model for Keyman Configuration, by refactoring the refresh of settings out of the render phase and more appropriately splitting applying and refreshing. It adds a new message flag for `wm_keyman_refresh`, `KR_SETTINGS_CHANGED`, which uses a random token approach broadcast to all applications on the desktop to ensure that the refresh is received and multiple notifications for the same event can be appropriately coalesced.

Most of the files touched are stripping out the refresh flags from the render phase.

There is one small bug resolved at the same time with the options properties, which were not reset to default if they had been deleted from the registry prior to a refresh (see utilkeymanoption.pas).

It would possibly be better to split the `Apply` and "notify" operations, but that is a substantial change to the behaviour of the program which I am not willing to do during a beta phase, and would possibly impact other users of the Keyman API.

# User testing

@MakaraSok:

For each operation, watch to see that:

- [ ] the keyboard menu is updated.
- [ ] the On Screen Toolbar is updated.
- [ ] Keyman Configuration reflects the changed settings accurately.

Try the following:

1. Installing a package from File Explorer while Keyman Configuration is open
2. Install a package from Keyman COnfiguration
3. Download a package from Keyman Configuration
4. Uninstall package
5. Enable and disable a keyboard
6. Adding and remove keyboard languages
7. Change various options in the Options tab
8. Change keyboard hotkeys

### CODE REVIEW

**Comment by MakaraSok** [comment]
All tests were done on Keyman 14.0.214-beta on Windows 10 Pro 20H2.

> 1. Installing a package from File Explorer while Keyman Configuration is open

**_Not working as expected_**

The Cameron keyboard package was installed from file while the Configuration is left opened. The Cameron keyboard is visible on keyboard menu on the Windows taskbar, but not in the Keyboard Layouts of the Configuration. 

image

> 2. Install a package from Keyman COnfiguration

**_Works as expected_**

Lao Basic package was installed from the Configuration using "Install keyboard...". The keyboard shows up in both Keyboard Layouts of the Configuration and Windows taskbar once the package is successfully installed.

> 3. Download a package from Keyman Configuration

**_Works as expected_**

CreeBoard package was downloaded and installed from the Configuration and as soon as the installed is finished, it shows up in the Keyboard Layouts.

> 4. Uninstall package

**_Works as expected_**

Kmhmu (SIL) keyboard disappears as soon as the removal process is done.

> 5. Enable and disable a keyboard

**_Works as expected_**

Lao Basic keyboard is unchecked in the Keyboard Layout and it disappears from the keyboard menu on the Windows taskbar.

> 6. Adding and remove keyboard languages

**_Works as expected_**

> 7. Change various options in the Options tab

**_Works as expected_**

Tested options:
- [not work] Uncheck the 'Keyboard hotkeys toggle keyboard activation,' the keyboard hotkey still works while it shouldn't. https://github.com/keymanapp/keyman/issues/4188
- [works] Check/uncheck 'Show hint messages'
- [works] Check/uncheck 'Simulate AltGr with Ctrl+Alt'
- [works] Check/uncheck 'Show splashscreen'
- [works] Check/uncheck 'Start when Windows starts'

_Other options have not been tested. I need more help on verifying whether or not those are working and not working._

> 8. Change keyboard hotkeys

There is no consistent result on this yet. Further investigation is needed as on one user the hotkey can be set and used with an issue while another seems to have no hotkey worked as expected.

This is from the user where hotkeys do not work. The key combo can be set, but not working at all. Look blank, but not really blank under the hood. See the screenshot below.

image

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas [hunk 1]
```diff
@@ -513,8 +513,10 @@ function TfrmInstallKeyboard.InstallTipForKeyboard(const BCP47Tag: string): Bool
 begin
   // Ensure keyboard is recorded in CU registry before we install the TIP, otherwise it will be marked as disabled by default,
   // as installing the TIP adds CU registry settings, potentially confusing the keyboard settings
-  kmcom.Refresh;
-  kmcom.Apply;
+
+  // Only do this for the keyboards, so we don't cause a global refresh
+  kmcom.Keyboards.Refresh;
+  kmcom.Keyboards.Apply;
```
Comment (jahorton) [inline windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:519]: I see a lot of places with `kmcom.Refresh` and `kmcom.Apply` in the new code... and a lot of places with `kmcom.Keyboards.Refresh` and `kmcom.Keyboards.Apply`.  Some of the places use half of each pair.

What determines which variant is used in a given circumstance?

File: windows/src/desktop/kmshell/install/uninstall.pas [hunk 2]
```diff
@@ -99,7 +99,7 @@ function UninstallPackage(Owner: TWinControl; const PackageName: WideString; Sil
     pkg := nil;
     if WaitForElevatedConfiguration(Handle, '-up "'+pkgID+'" -s') = 0 then
     begin
-      kmcom.Keyboards.Refresh;
+      kmcom.Refresh;
       kmcom.Keyboards.Apply;
```
Comment (jahorton) [inline windows/src/desktop/kmshell/install/uninstall.pas:103]: For example, why do we use `kmcom.Refresh` but not `kmcom.Apply` here, instead using `kmcom.Keyboards.Apply`?

File: windows/src/desktop/kmshell/install/uninstall.pas [hunk 3]
```diff
@@ -165,7 +165,7 @@ function UninstallKeyboard(Owner: TWinControl; const KeyboardName: WideString; S
   kbd.Uninstall;
   kbd := nil;
   kmcom.Keyboards.Refresh;
-  kmcom.Keyboards.Apply;
+  kmcom.Apply;
```
Comment (jahorton) [inline windows/src/desktop/kmshell/install/uninstall.pas:168]: Yet here, we use `kmcom.Apply` over `kmcom.Keyboards.Apply`, despite using `kmcom.Keyboards.Refresh`?

**Review by jahorton — COMMENTED** [review]
Unfortunately, there's one aspect of the changes that is anything but "simple and clear" to review, and it pops up in a _lot_ of the changed sections.

**Comment by jahorton** [comment]
Note that I haven't exhaustively listed out each confusing case; I just picked some of the earliest examples.

File: windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas (hunk 1 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/UfrmInstallKeyboard.pas:519]: This is the result of an unfortunate API side-effect. `kmcom.Apply` not only writes the changes out, it also notifies all apps on the system of the changes. However, `kmcom.Keyboards.Apply` only writes the changes for the keyboards and does not do the global notification. So when we are midway through a process, often we'll just `Apply` the area we know has changed, to reduce chatter and unnecessary multiple notifications.

File: windows/src/desktop/kmshell/install/uninstall.pas (hunk 2 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/uninstall.pas:103]: This should probably be `kmcom.Apply`, matching the code changes further down. I think I just missed it.

```suggestion
      kmcom.Refresh;
      kmcom.Apply;
```

File: windows/src/desktop/kmshell/install/uninstall.pas (hunk 3 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/uninstall.pas:168]: Tricksy, isn't it. We can do a full `kmcom.Refresh` / `kmcom.Apply` here, which would be better. See also lines 157-158 which should really be the same. I'll fix those up.

File: windows/src/desktop/kmshell/install/uninstall.pas [hunk 4]
```diff
@@ -199,6 +199,7 @@ function UninstallKeyboardLanguage(const LanguageProfileGUID: WideString; Silent
   kbdlang.Uninstall;
   kbdlang := nil;
   kmcom.Keyboards.Refresh;
```
Comment (mcdurdin) [inline windows/src/desktop/kmshell/install/uninstall.pas:201]:
```suggestion
  kmcom.Refresh;
```

File: windows/src/desktop/kmshell/install/uninstall.pas (hunk 3 above)
Comment (mcdurdin) [reply windows/src/desktop/kmshell/install/uninstall.pas:168]: Eventually, I want to refactor all this to stop using a global `kmcom` and instantiate the class locally at each change. But that's a much bigger job with potential performance side-effects and mucho testing required, so we'll live with some of the unnecessary stuff such as the `Refresh`+`Apply` pattern which probably only needs to be an `Apply` pattern when making changes on a local `kmcom` object and its members.

**Comment by mcdurdin** [comment]
@MakaraSok did you test on the PR build or on the official beta build, because these changes haven't been merged yet.

I suggest that you test on the PR build for #4193 because that and a number of other PRs build on this one.

**Review by jahorton — APPROVED** [review]

**Comment by mcdurdin** [comment]
@MakaraSok can you retest on 14.0.215-beta tomorrow?

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.215-beta](https://keyman.com/downloads/releases/beta/14.0.215)

**Comment by MakaraSok** [comment]
Tested with v14.0.221-beta. The followings work as expected:

> the keyboard menu is updated.
> the On Screen Toolbar is updated.
> Keyman Configuration reflects the changed settings accurately.

Except: #4289 and #4290.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)

**Pull request merged**